### ISSUE TITLE
API Versioning?

### ISSUE THREAD

**Issue by dlandi** [stratisproject/StratisBitcoinFullNode#1316]
Hi.

While the FullNode is in beta, this might be a good time to bring up the issue of versioning.  Have you considered how to introduce new versions of existing features?  It may well be that legacy versions will need to run alongside newer versions.  The MVC Core 2 framework provides a versioning interface. There is more than one way to do this. But first you need to install the Versioning nuget package:

Install-Package Microsoft.AspNetCore.Mvc.Versioning 

And then you need to add it to the StartUp.ServiceCollection.

public void ConfigureServices(IServiceCollection services ) 
{ 
	services.AddMvc();     
	services.AddApiVersioning(); 
}

Here is an example of reworking the Notifications Controller two support multiple versions.

----------------------------------------------------------------------------------------

using System.Net;
using Microsoft.AspNetCore.Mvc;
...
using Microsoft.AspNetCore.Mvc.Versioning;

------------------------------ Version One ---------------------------------------
namespace Stratis.Bitcoin.Features.Notifications.Controllers
{
    /// <summary>
	//	Version 1 of NotificationsController
    /// Controller providing operations on blocks and transactions notifications.
    /// </summary>
    [Route("api/[controller]")]
	[ApiVersion("1.0")]
    public class NotificationsController : Controller
    {
		...
		
        [HttpGet]
        [Route("sync")]
        public IActionResult SyncFrom([FromQuery] string from)
        {
            //....
            return this.Ok();
        }
    }

------------------------------ Version Two ---------------------------------------
	
    /// <summary>
	//	Version 2 of NotificationsController	
    /// Controller providing operations on blocks and transactions notifications.
    /// </summary>
    [Route("api/[controller]")]
	[ApiVersion("2.0")]
    public class NotificationsController : Controller
    {
        ...

        [HttpGet]
        [Route("sync")]
        public IActionResult SyncFrom([FromQuery] string from)
        {
            //....
            return this.Ok();
        }
    }
	
}

-----------------------------

Usage: 

/api/notification/sync?from=foo&version=1.0
/api/notification/sync?from=bar&version=2.0

By simply adding the Versioning nuget package, you leave this option open to your features in the future.

Just a thought.

-dennis

**Comment by dangershony** [comment]
Great idea, its more about versioning the feature endpoints and not the features themselves.
Features already have versioning built in as part of nuget but we don't enforce feature version dependency

**Comment by dlandi** [comment]
A good article on API Versioning:
https://neelbhatt.com/2018/04/21/api-versioning-in-net-core/

**Comment by sondreb** [comment]
I have made a prototype for versioning of the API, I decided to go with number-versioning as oppose to date-numbering, which has become a standard for many, especially Microsoft. I do think number version is the best fit for this API.

I went with query-versioning, with the query named "api-version" (becoming more or less industry standard), and not versions in header nor URL.

Started by adding some new methods to the API, and then I realized the existing API could be improved somewhat to be more "restyfied", and instead of making to much fuzz with existing API, I added versioning to keep them separate.

The idea with the new BlockStore API, is that /api/blocks should return the 10 latest blocks, always as JSON (not have an option to return just the hashes). Then /api/blocks/page/2 should return the 20 first blocks of the blockchain. This would make those page URLs propper resources, meaning they do not change on subsequent requests. I might consider adding a page size as query parameter, but think it should rather stick to a fixed one of 10.

Is this something that would be of interest to have a PR on when I'm potentially closer to some stable API?

Here is a screenshot of "v2.0":

![image](https://user-images.githubusercontent.com/309938/44289646-6e49be00-a275-11e8-9f45-c5b50c248e41.png)
